Authentication requirements in MetalK8s

thomasdanan opened the pull request:
**Component**:

Requirements doc

**Context**: 

Requirements about Authentication in MetalK8s

**Summary**:

Listing context, requirements and user story

**Acceptance criteria**: 

Agree on requirements with contributors

thomasdanan committed:
Remove eveything related to authorization and focus the spec on authentication only ...

thomasdanan committed:
Add requirement about login page theming and authentication check from the UI

bert-e commented:
# Hello thomasdanan,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort

NicolasT commented on docs/developer/architecture/authentication.rst:
```diff
@@ -0,0 +1,88 @@
+Context:
+--------
+Currently, when we deploy MetalK8s we pre provision super admin user with admin
+password. Meaning anyone wanting to use K8S/Salt APIs should authenticate using
```
One can trivially add more users as well, with whatever group mappings, and set up custom `(Cluster)RoleBinding`s, so this isn't entirely true.

thomasdanan replied:
updated in last commit

NicolasT commented on docs/developer/architecture/authentication.rst:
```diff
@@ -0,0 +1,88 @@
+this single super admin user.
+This is not good enough in term of security and flexibility.
+
+We would like at least to be able to add different users with different
```
As said above, you *can* do that today (though there are security aspects in the current solution that can't be ignored).

thomasdanan replied:
updated

NicolasT commented on docs/developer/architecture/authentication.rst:
```diff
@@ -0,0 +1,88 @@
+- Being able to provision users with an embedded identity provider
+- Being able to integrate with an external Identity provider
+
+Dex https://github.com/dexidp/dex seems to be the preferred solution for now,
```
Let's write proper ReST documents with real links and such ;-)

NicolasT reviewed (dismissed):
WiP

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort





The following reviewers are expecting changes from the author, or must review again:

* @nicolast

NicolasT commented on docs/developer/architecture/authentication.rst:
```diff
@@ -0,0 +1,88 @@
+
+**Dex as Identity Provider and login page**
+
+Dex login page theme should be as closed as possible as the rest of UIs themes.
```
```suggestion
Dex login page theme should be as close as possible as the rest of UIs themes.
```

thomasdanan committed:
update following spec review
- explicitely add the need for a cli tool doing user provisioning
- explicitely add the need for a cli tool doing external IDP integration
- add req about SSO
- add req about integrating customer tools to our APIs
- more precise about how DEX login page should behave

thomasdanan committed:
fix rst formatting issues

Ebaneck commented on docs/developer/architecture/authentication.rst:
```diff
@@ -0,0 +1,97 @@
+-------------
+**User Management with Dex as an IDP**
+
+As an IT Generalist, I want to provision/edit usersa and high level roles.
```
_Minor typos_
As an IT Generalist, I want to provision/edit **users** and high-level roles.

thomasdanan replied:
done

gdemonet commented on docs/developer/architecture/authentication.rst:
```diff
@@ -0,0 +1,97 @@
+Alert Manager instances. Some other may want to leverage K8S / Salt APIs to
+drive administration of the system. Those APIs should be secured enough either
+through OIDC, Service account or because only accessible through control plane
+Network.
```
So, what's the actual requirement? Is being accessible only in the CP network sufficient? If so, then it's done already.

thomasdanan replied:
moved it to open question

gdemonet commented on docs/developer/architecture/authentication.rst:
```diff
@@ -0,0 +1,97 @@
+authenticating with DEX local users or with external IDP.
+If no external IDP is configured, no choice is presented to the user.
+This DEX login page should be styled so that it looks like any other Scality
+web pages.
```
Would make sense to mention that all Admin UIs should rely on this same authentication mechanism

thomasdanan replied:
done

gdemonet commented on docs/developer/architecture/authentication.rst:
```diff
@@ -0,0 +1,97 @@
+As an IT Generalist, I want to leverage my IDP in order to reuse already
+provisioned users & groups.
+The way we do that integration is through a CLI tool. This CLI tool will take
+LDAP or AD integration configuration file as an input.
```
I'd argue this is not in the realm of "requirements", but an implementation proposal.

Requirements could be summed up like:
- Users want to be able to configure external identity providers,
- they want to be able to do so without specific K8s/Dex knowledge

To which an implementation design would suggest developing a CLI tool to abstract K8s/Dex details.

thomasdanan replied:
done

gdemonet commented on docs/developer/architecture/authentication.rst:
```diff
@@ -0,0 +1,97 @@
+- Being able to integrate with an external IDP
+
+`Dex <https://github.com/dexidp/dex>`_ seems to be the preferred solution for
+now, for both local identity provider or proxy to external IDP.
```
Mentions of `Dex` in the requirements seem inappropriate. To me, `Dex` is _one_ solution to the set of requirements, and this choice should be documented in the "design" section of this architecture document, along with the discarded alternatives.

thomasdanan replied:
ok so will replace DEX with local IDP

gdemonet commented on docs/developer/architecture/authentication.rst:
```diff
@@ -0,0 +1,97 @@
+The CLI tool enables to add / delete and edit passwords and roles.
+
+The CLI tool is used right after bootstrap node deployment to provision the
+very first user (which won't be provisioned anymore by default)
```
That's a really good point, but is "lockdown by default" really a user requirement?

thomasdanan replied:
Actually I changed my mind after discussing with other folks ... we want one pre provisioned admin and we want to ask him to change the default password at very first connection ... I put back the original requirement.

gdemonet commented on docs/developer/architecture/authentication.rst:
```diff
@@ -0,0 +1,97 @@
+---------------
+
+- Authentication across multiple sites
+- SSO across Admin UIs (control plane) and Zenko Management UIs (on data plane)
```
Let's not mention other products in MetalK8s documents, or only as examples. In here, the question would be "Unified authentication and SSO between Admin UIs and other workload UIs".

Since those lie in different networks and namespaces, it's probably not achievable - but discussing this point would go in the "design" section of this document, IMHO.

thomasdanan replied:
done

Ebaneck commented on docs/developer/architecture/authentication.rst:
```diff
@@ -0,0 +1,97 @@
+The CLI tool enables to add / delete and edit passwords and roles.
+
+The CLI tool is used right after bootstrap node deployment to provision the
+very first user (which won't be provisioned anymore by default)
```
I would argue that this will not be the first user to be provisioned.

I still think that the `Scality-Dex` superuser needs to be statically defined else we will have no way to connect Dex to the API-server.
Adding to this, the user experience becomes even more awful when we have to ask users to configure the API-server/Dex integration and force reboot the API-server.
https://github.com/dexidp/dex/blob/master/Documentation/kubernetes.md#configuring-the-openid-connect-plugin

gdemonet replied:
No, you don't need a user to have K8s API accept OIDC tokens issued from Dex. As explained in your docs, the setup requires certificates to be generated (that's going to be done with Salt and the `ca:minion`), said certs must then be exposed to API Server, and that should be enough.

NicolasT replied:
Dex connecting to K8s API is using a `ServiceAccount`, so no need to provision any OIDC user. And as @gdemonet said, you can set up K8s API to accept OIDC tokens *even* when the OIDC service is not (yet) running. No need for any restart.

Which is a good thing, otherwise your whole cluster would stop operating when there's an issue with some external component, i.e. the OIDC IdP in this case?

To define new users (in the case of Dex), we could either use its gRPC API (though I have some concerns about authn towards said API...), or work at the `ConfigMap` level, which is not an issue since we *have* `/etc/kubernetes/admin.conf` available when the user management script is executed. No need for any OIDC 'user' either.

Ebaneck commented on docs/developer/architecture/authentication.rst:
```diff
@@ -0,0 +1,97 @@
+
+Once DEX is in place and user are provisioned through external IDP one
+authenticated user can easily navigate to the other admin UIs without having
+to re-authenticate.
```
Once DEX is in place and **users** are provisioned through external IDP, authenticated users can easily navigate......
??

thomasdanan replied:
agree, this was confusing ... wanted to say whatever IDP is in place ... will fix

Ebaneck reviewed (changes requested):
A couple of Minors to fix

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort





The following reviewers are expecting changes from the author, or must review again:

* @nicolast

* @ebaneck

NicolasT commented on docs/developer/architecture/authentication.rst:
```diff
@@ -0,0 +1,97 @@
+The CLI tool enables to add / delete and edit passwords and roles.
+
+The CLI tool is used right after bootstrap node deployment to provision the
+very first user (which won't be provisioned anymore by default)
```
There's another req that should be added, IMHO, about the Vagrant deployment.

Similar to how we create (sparse) volumes for Prometheus and such by default when running `vagrant up`, I think it could make sense to also pre-provision some well-known user in said environment to retain a cluster that's fully working out of the box.

thomasdanan replied:
Actually I changed this req and put back the need of having a pre-provisioned user so I guess it would not be different with vagrant N

bert-e commented:
# Branches have diverged



This pull request's source branch `feature/authentication` has diverged from
`development/2.4` by more than 50 commits.

To avoid any integration risks, please re-synchronize them using one of the
following solutions:

* Merge `origin/development/2.4` into `feature/authentication`
* Rebase `feature/authentication` onto `origin/development/2.4`

Note: If you choose to rebase, you may have to ask me to rebuild
integration branches using the `reset` command.

thomasdanan committed:
few fixes following PR reviex:
- Replace DEX with local IDP
- Re introduce the concept of pre provisioned user
- Fix typos

thomasdanan commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Branches have diverged



This pull request's source branch `feature/authentication` has diverged from
`development/2.4` by more than 50 commits.

To avoid any integration risks, please re-synchronize them using one of the
following solutions:

* Merge `origin/development/2.4` into `feature/authentication`
* Rebase `feature/authentication` onto `origin/development/2.4`

Note: If you choose to rebase, you may have to ask me to rebuild
integration branches using the `reset` command.

thomasdanan committed:
remove prom / am secured requirement
but keep it as an open question

thomasdanan commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Branches have diverged



This pull request's source branch `feature/authentication` has diverged from
`development/2.4` by more than 50 commits.

To avoid any integration risks, please re-synchronize them using one of the
following solutions:

* Merge `origin/development/2.4` into `feature/authentication`
* Rebase `feature/authentication` onto `origin/development/2.4`

Note: If you choose to rebase, you may have to ask me to rebuild
integration branches using the `reset` command.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort





The following reviewers are expecting changes from the author, or must review again:

* @nicolast

* @ebaneck

thomasdanan commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort





The following reviewers are expecting changes from the author, or must review again:

* @nicolast

* @ebaneck

Ebaneck committed:
- Minor restructuring of sentences
- Add authentication document to toctree

gdemonet commented on docs/developer/architecture/authentication.rst:
```diff
@@ -0,0 +1,104 @@
+- Our customers may want to collect some statistics out of our Prometheus
+  instances. This API should be secured enough and could be made available
+  either through OIDC, Service account tokens or be made accessible only
+  through the control plane network.
```
This sentence could be reformulated to include the known considerations, e.g.:

```
This API could be authenticated using OIDC, using an OIDC proxy, or stay unauthenticated. One should consider:
- the low sensitivity of the exposed data
- the fact that it is only exposed on the control-plane network
- the fact that most consumers of Prometheus stats are not human (e.g. Grafana, a federating Prometheus, scripts...), hence not well-suited to performing the OIDC dance
```

Ebaneck reviewed (approved):
Let's merge it

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort





The following reviewers are expecting changes from the author, or must review again:

* @nicolast

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort

thomasdanan commented:
/approve

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.4`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.4`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`



Please check the status of the associated issue None.

Goodbye thomasdanan.
